yapapi/runner/events.py

```diff
@@ -1,161 +1,146 @@
+    DownloadStarted = NamedTuple("DownloadStarted", [("path", str)])
+    DownloadFinished = NamedTuple("DownloadFinished", [("path", str)])
 
-    DOWNLOAD_STARTED = auto()
-    DOWNLOAD_FINISHED = auto()
+    def __init__(self, *_args, **_kwargs):
+        raise NotImplementedError(f"Cannot create instances of {type(self).__name__}")
 
 
 EventType = Union[
-    SubscriptionEvent, ProposalEvent, AgreementEvent, WorkerEvent, TaskEvent, StorageEvent
+    Event.ComputationStarted,
+    Event.ComputationFinished,
+    Event.ComputationFailed,
+    Event.SubscriptionCreated,
+    Event.SubscriptionFailed,
+    Event.CollectFailed,
+    Event.ProposalFailed,
+    Event.ProposalReceived,
+    Event.ProposalRejected,
+    Event.ProposalResponded,
+    Event.ProposalConfirmed,
+    Event.AgreementCreated,
+    Event.AgreementConfirmed,
+    Event.AgreementRejected,
+    Event.PaymentAccepted,
+    Event.PaymentPrepared,
+    Event.PaymentQueued,
+    Event.InvoiceReceived,
+    Event.WorkerStarted,
+    Event.ActivityCreated,
+    Event.ActivityCreateFailed,
+    Event.TaskStarted,
+    Event.WorkerFinished,
+    Event.ScriptSent,
+    Event.CommandExecuted,
+    Event.GettingResults,
+    Event.ScriptFinished,
+    Event.TaskAccepted,
+    Event.TaskRejected,
+    Event.DownloadStarted,
+    Event.DownloadFinished,
 ]
 
 
-_event_type_to_string = {
-    SubscriptionEvent.CREATED: "Proposal subscription created",
-    SubscriptionEvent.FAILED: "Failed to subscribe to proposals",
-    SubscriptionEvent.COLLECT_FAILED: "Failed to collect proposals",
-    ProposalEvent.BUFFERED: "Proposal buffered",
-    ProposalEvent.FAILED: "Proposal failed",
-    ProposalEvent.RECEIVED: "Proposal received",
-    ProposalEvent.REJECTED: "Proposal rejected",
-    ProposalEvent.RESPONDED: "Responded to proposal",
-    AgreementEvent.CREATED: "Agreement created",
-    AgreementEvent.CONFIRMED: "Agreement confirmed",
-    AgreementEvent.REJECTED: "Agreement rejected",
-    AgreementEvent.PAYMENT_ACCEPTED: "Payment accepted for agreement",
-    AgreementEvent.PAYMENT_PREPARED: "Payment prepared for agreement",
-    AgreementEvent.PAYMENT_QUEUED: "Payment queued for agreement",
-    AgreementEvent.INVOICE_RECEIVED: "Invoice received for agreement",
-    WorkerEvent.CREATED: "Worker created",
-    WorkerEvent.ACTIVITY_CREATED: "Activity created",
-    WorkerEvent.ACTIVITY_CREATE_FAILED: "Failed to create activity",
-    WorkerEvent.GOT_TASK: "Worker got new task",
-    WorkerEvent.FINISHED: "Worker has finished",
-    TaskEvent.SCRIPT_SENT: "Script sent to provider",
-    TaskEvent.COMMAND_EXECUTED: "Command executed",
-    TaskEvent.GETTING_RESULTS: "Getting task results",
-    TaskEvent.SCRIPT_FINISHED: "Script finished",
-    TaskEvent.ACCEPTED: "Task accepted",
-    TaskEvent.REJECTED: "Task rejected",
-    StorageEvent.DOWNLOAD_STARTED: "Download started",
-    StorageEvent.DOWNLOAD_FINISHED: "Download finished",
+if get_type_args:
```

Comment:
Clever; you can also check if `event_type_to_string` contains all human-readable representations.

Reply:
why `Event` is not just a superclass for these types. I would avoid a Union greater than 4 components

Reply:
Originally, I wanted the definition of events to be as concise as possible, preferably with each event type definition in one line (they were elements of an enum previously). But as you noticed this had its drawbacks, for example the need for this big `Union` type. So I finally changed this so that each event type is a dataclass inheriting from `Event`. No need for `Union` now!